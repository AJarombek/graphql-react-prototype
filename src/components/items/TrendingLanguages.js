/**
 * TrendingLanguages component which displays top languages in the past two years.
 * @author Andrew Jarombek
 * @since 4/21/2020
 */

import React, { useEffect, useState } from 'react';
import moment from 'moment';
import { getRecentTopLanguages } from '../../datasource/GraphQL';

const TrendingLanguages = () => {
  const [trendingLanguages, setTrendingLanguages] = useState([]);
  const [error, setError] = useState(null);

  useEffect(() => {
    getRecentTopLanguages('AJarombek')
      .then(result => {
        if (result.data.data) {
          generateTrendingLanguages(result.data.data.user.repositories.edges);
        } else {
          setError(result.data.errors[0].message);
        }
      });
  }, []);

  const generateTrendingLanguages = (repositories) => {
    const languages = {};

    for (let repo of repositories) {
      const creationDate = repo.node.createdAt;

      if (moment(creationDate) >= moment("2019-01-01")) {
        const language = repo.node.primaryLanguage.name;

        if (languages.hasOwnProperty(language)) {
          languages[language] = languages[language] + 1;
        } else {
          languages[language] = 1;
        }
      } else {
        break;
      }
    }

    const sortedLanguages = [];
    for (const language in languages) {
      sortedLanguages.push({
        name: language,
        occurrences: languages[language]
      });
    }

    sortedLanguages.sort((a, b) => b.occurrences - a.occurrences);

    setTrendingLanguages(sortedLanguages.slice(0, 5));
    setError(null);
  };

  return (
    <div className="items trending-languages">
      {error ?
        <div className="error">
          <h6>{error}</h6>
        </div>
        :
        <>
          <h2>Trending Languages</h2>
          { trendingLanguages.map(language =>
            <>
              <p>{language.name}</p>
              <p>{language.occurrences}</p>
            </>
          )
          }
        </>
      }
    </div>
  );
};

export default TrendingLanguages;
