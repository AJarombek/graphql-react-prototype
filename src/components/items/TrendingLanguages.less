/**
 * Stylesheet for the TrendingLanguages component.
 * Author: Andrew Jarombek
 * Date: 4/21/2020
 */

@import "../../vars";

.trending-languages {
  grid-area: trndLang;

  h2 {
    margin-bottom: 20px;
  }

  .language:nth-child(2) {
    p:nth-child(1) {
      font-size: 20px;
    }

    p:nth-child(2) {
      font-size: 20px;
      .hoverPurple();
    }
  }

  .language {
    display: flex;
    align-items: center;
    margin: 8px 0;
    width: 250px;

    p:nth-child(1) {
      font-size: 16px;
    }

    p:nth-child(2) {
      font-size: 18px;
      margin: 0 0 0 auto;
    }
  }
}
