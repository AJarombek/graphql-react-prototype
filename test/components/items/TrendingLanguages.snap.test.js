/**
 * Snapshot test for the TrendingLanguages component.
 * @author Andrew Jarombek
 * @since 4/25/2020
 */

import React from 'react';
import TrendingLanguages from '../../../src/components/items/TrendingLanguages';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(<TrendingLanguages />).toJSON();
  expect(tree).toMatchSnapshot();
});
