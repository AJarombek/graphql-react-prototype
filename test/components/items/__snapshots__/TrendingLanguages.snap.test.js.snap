// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="items trending-languages"
>
  <h2>
    Trending Languages
  </h2>
</div>
`;
